firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/Sensor.cpp

```diff
@@ -14,65 +14,80 @@ namespace Pufferfish::Driver::Serial::Nonin {
 // Sensor
 
 InitializableState Sensor::setup() {
-  // sensor is connected when we have measurements available
-  switch (measure(measurements_)) {
-    case InitializableState::failed:
+  switch (device_.output(measurements_)) {
+    case PacketStatus::invalid_checksum:
+    case PacketStatus::invalid_header:
       return InitializableState::failed;
-    case InitializableState::ok:
-      sensor_connections_.nonin_connected = true;
-      return InitializableState::ok;
-    case InitializableState::setup:
+    case PacketStatus::waiting:
+    case PacketStatus::frame_loss:
+    case PacketStatus::ok:
+      input_clock(time_.millis());
       break;
   }
 
+  if (wait_time_exceeded()) {
+    return InitializableState::failed;
+  }
+
   return InitializableState::setup;
 }
 
 InitializableState Sensor::output(float &spo2, float &hr) {
-  if (!sensor_connections_.nonin_connected) {
+  switch (device_.output(measurements_)) {
+    case PacketStatus::invalid_checksum:
+    case PacketStatus::invalid_header:
+      // handle error cases first
+      spo2 = NAN;
+      hr = NAN;
+      return InitializableState::failed;
+    case PacketStatus::waiting:
+    case PacketStatus::frame_loss:
+    case PacketStatus::ok:
```

Comment:
`ok` seems like a special case to handle differently from the other cases. (TBD: same with setup method?)

Reply:
`ok` should reset the waiting timeout timer. (same with setup?)